llvm/test/Transforms/Util/loop-iter-count-asserts.ll

```diff
@@ -0,0 +1,298 @@
+; CHECK:       if.end.i.i.i.i:
+; CHECK-NEXT:    br label [[FOR_BODY_EXIT]]
+; CHECK:       for.body.exit:
+; CHECK-NEXT:    [[ADD]] = add nuw nsw i32 [[I_014]], 1
+; CHECK-NEXT:    [[CMP2:%.*]] = icmp ult i32 [[ADD]], [[DIV7]]
+; CHECK-NEXT:    br i1 [[CMP2]], label [[FOR_BODY]], label [[FOR_COND_CLEANUP:%.*]], !llvm.loop [[LOOP0]]
+;
+entry:
+  %0 = load i32, ptr %params, align 32
+  %div7 = lshr i32 %0, 6
+  br label %for.body
+
+for.cond.cleanup:
+  ret void
+
+for.body:
+  %i.014 = phi i32 [ 0, %entry ], [ %add, %for.body.exit ]
+  %op_params.i.i.i = getelementptr inbounds i8, ptr %params, i20 32
+  %3 = load i8, ptr %op_params.i.i.i, align 32
+  %cmp3 = icmp eq i8 %3, -1
+  %4 = tail call i1 @llvm.is.constant.i1(i1 %cmp3)
+  %5 = and i1 %cmp3, %4
+  br i1 %5, label %for.body.exit, label %if.end.i.i.i.i
+
+if.end.i.i.i.i:
+  br label %for.body.exit
+
+for.body.exit:
+  %add = add nuw nsw i32 %i.014, 1
+  %cmp2 = icmp ult i32 %add, %div7
+  br i1 %cmp2, label %for.body, label %for.cond.cleanup, !llvm.loop !6
+}
+
+; Header has unconditional branch, abort Loop rotataion and metadata adding
+; Function Attrs: mustprogress noinline
+define  dso_local void @header_unconditional(ptr nonnull align 32 dereferenceable(64) %params) {
+; CHECK-LABEL: @header_unconditional(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = load i32, ptr [[PARAMS:%.*]], align 32
+; CHECK-NEXT:    [[DIV7:%.*]] = lshr i32 [[TMP0]], 6
+; CHECK-NEXT:    [[CMP211:%.*]] = icmp ult i32 0, [[DIV7]]
+; CHECK-NEXT:    br i1 [[CMP211]], label [[FOR_BODY_PREHEADER:%.*]], label [[FOR_COND_CLEANUP:%.*]]
+; CHECK:       for.body.preheader:
+; CHECK-NEXT:    br label [[FOR_BODY:%.*]]
+; CHECK:       for.cond.cleanup.loopexit:
+; CHECK-NEXT:    br label [[FOR_COND_CLEANUP]]
+; CHECK:       for.cond.cleanup:
+; CHECK-NEXT:    ret void
+; CHECK:       for.body:
+; CHECK-NEXT:    [[I_014:%.*]] = phi i32 [ [[ADD:%.*]], [[FOR_BODY_EXIT:%.*]] ], [ 0, [[FOR_BODY_PREHEADER]] ]
+; CHECK-NEXT:    br label [[FOR_BODY_EXIT]]
+; CHECK:       for.body.exit:
+; CHECK-NEXT:    [[ADD]] = add nuw nsw i32 [[I_014]], 1
+; CHECK-NEXT:    [[CMP2:%.*]] = icmp ult i32 [[ADD]], [[DIV7]]
+; CHECK-NEXT:    br i1 [[CMP2]], label [[FOR_BODY]], label [[FOR_COND_CLEANUP_LOOPEXIT:%.*]], !llvm.loop [[LOOP0]]
+;
+entry:
+  %0 = load i32, ptr %params, align 32
+  %div7 = lshr i32 %0, 6
+  %cmp211 = icmp ult i32 0, %div7
+  br i1 %cmp211, label %for.body, label %for.cond.cleanup
+
+for.cond.cleanup:
+  ret void
+
+for.body:
+  %i.014 = phi i32 [ 0, %entry ], [ %add, %for.body.exit ]
+  br label %for.body.exit
+
+for.body.exit:
+  %add = add nuw nsw i32 %i.014, 1
+  %cmp2 = icmp ult i32 %add, %div7
+  br i1 %cmp2, label %for.body, label %for.cond.cleanup, !llvm.loop !6
+}
+
+
+!3 = distinct !{!3, !11, !7, !9}
+!6 = distinct !{!6, !7, !8, !9}
+!7 = !{!"llvm.loop.mustprogress"}
+!8 = !{!"llvm.loop.itercount.range", i64 4}
+!9 = !{!"llvm.loop.unroll.disable"}
+!11 = !{!"llvm.loop.itercount.range", i64 10, i64 65}
```

Comment:
I would like to see some more cases, e.g. decrementing to zero, using eq/ne, having a single do {} while block, ...

Reply:
I also added a loop guard test in addition to your suggestions.